src/radiation_system.hpp

```diff
@@ -611,264 +691,284 @@ void RadSystem<problem_t>::ComputeFluxes(array_t &x1Flux_in, array_t &x1FluxDiff
+			const double S_R = std::max(0.1 * c_hat_, c_hat_ * std::sqrt(Tnormal_R));
+
+			AMREX_ASSERT(std::abs(S_L) <= c_hat_); // NOLINT
+			AMREX_ASSERT(std::abs(S_R) <= c_hat_); // NOLINT
+
+			// in the frozen Eddington tensor approximation, we are always
+			// in the star region, so F = F_star
+			const quokka::valarray<double, numRadVars_> F =
+			    (S_R / (S_R - S_L)) * F_L - (S_L / (S_R - S_L)) * F_R + epsilon * (S_R * S_L / (S_R - S_L)) * (U_R - U_L);
+
+			// check states are valid
+			AMREX_ASSERT(!std::isnan(F[0])); // NOLINT
+			AMREX_ASSERT(!std::isnan(F[1])); // NOLINT
+			AMREX_ASSERT(!std::isnan(F[2])); // NOLINT
+			AMREX_ASSERT(!std::isnan(F[3])); // NOLINT
+
+			x1Flux(i, j, k, radEnergy_index + numRadVars_ * g - nstartHyperbolic_) = F[0];
+			x1Flux(i, j, k, x1RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = F[1];
+			x1Flux(i, j, k, x2RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = F[2];
+			x1Flux(i, j, k, x3RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = F[3];
+
+			const quokka::valarray<double, numRadVars_> diffusiveF =
+			    (S_R / (S_R - S_L)) * F_L - (S_L / (S_R - S_L)) * F_R + (S_R * S_L / (S_R - S_L)) * (U_R - U_L);
+
+			x1FluxDiffusive(i, j, k, radEnergy_index + numRadVars_ * g - nstartHyperbolic_) = diffusiveF[0];
+			x1FluxDiffusive(i, j, k, x1RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = diffusiveF[1];
+			x1FluxDiffusive(i, j, k, x2RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = diffusiveF[2];
+			x1FluxDiffusive(i, j, k, x3RadFlux_index + numRadVars_ * g - nstartHyperbolic_) = diffusiveF[3];
+		} // end loop over radiation groups
 	});
 }
 
-template <typename problem_t> AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputePlanckOpacity(const double /*rho*/, const double /*Tgas*/) -> double
+template <typename problem_t>
+AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputePlanckOpacity(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
 {
-	return NAN;
+	quokka::valarray<double, nGroups_> kappaPVec{};
+	for (int g = 0; g < nGroups_; ++g) {
+		kappaPVec[g] = NAN;
+	}
+	return kappaPVec;
 }
 
 template <typename problem_t>
-AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputePlanckOpacityTempDerivative(const double /*rho*/, const double /*Tgas*/) -> double
+AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputeFluxMeanOpacity(const double /*rho*/, const double /*Tgas*/) -> quokka::valarray<double, nGroups_>
 {
-	return 0.0;
+	quokka::valarray<double, nGroups_> kappaFVec{};
+	for (int g = 0; g < nGroups_; ++g) {
+		kappaFVec[g] = NAN;
+	}
+	return kappaFVec;
 }
 
-template <typename problem_t> AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputeRosselandOpacity(const double /*rho*/, const double /*Tgas*/) -> double
+// TODO(CCH): Derivative of Planck/energy-mean opacity with respect to gas temperature. In the future, they should be defined separately.
```

Comment:
I don't understand what you mean here. Can you explain this comment?

Reply:
Similar to my other comments about d B_i / d T_gas, this is something I plan to do next. Mark and I discussed about this, and our plan is to define a class to allow the user to specify d kappa_P / d Tgas. We will also provide a convenient function to compute this from the Integral of Planck function if the user provide a tabulated kappa(nu). I removed this TODO and kept the other one about d B_i / d T_gas. 

I removed all TODOs except a few of them which I find it necessary to keep.